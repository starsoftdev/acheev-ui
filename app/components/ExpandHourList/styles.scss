@import '~styles/settings/settings';

.expandHourList {
  font-size: $fs-sm;
  text-transform: uppercase;

  &__labelWrapper {
    cursor: pointer;
    font-weight: bold;
  }

  &__labelWrapper,
  &__dropdownWrapper {
    display: flex;
    align-items: center;
  }

  &__label {
    width: rem-calc(50);
    padding-right: $rem-sm;
  }
  &__status {
    width: rem-calc(95);
    padding-right: $rem-sm;
    text-align: center;
    &--danger {
      color: $danger;
    }

    &--success {
      color: $success;
    }

    &--active {
      font-weight: bold;
    }
  }
  &__time {
    padding-right: $rem-sm;
    min-width: rem-calc(125);
    text-align: center;
  }
}
