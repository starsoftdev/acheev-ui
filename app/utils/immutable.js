// @flow

import { Iterable } from 'immutable';
import type { Map } from 'immutable';
import { get as _get } from 'lodash-es';

export const get = (
  collection: ?Map<*, *> | Object,
  path: string | number,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.get(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};

export const getIn = (
  collection: ?Map<*, *> | Object,
  path: Array<string | number>,
  defaultValue: any
) => {
  if (!collection) {
    return defaultValue;
  } else if (Iterable.isIterable(collection)) {
    return collection.getIn(path, defaultValue);
  }
  return _get(collection, path, defaultValue);
};
