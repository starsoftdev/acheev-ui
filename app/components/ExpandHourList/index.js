// @flow

import * as React from 'react';
import cx from 'classnames';
import moment from 'moment-timezone';
import Icon from 'components/Icon';

import ChevronDown from 'images/sprite/chevron-down.svg';
import ChevronUp from 'images/sprite/chevron-up.svg';
import './styles.scss';

const weekNames = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
  'Sunday',
];

type Props = {
  data: Object,
  date?: string,
};

type State = {
  show: boolean,
};

const getTimeString = data => {
  if (data.get('isActive') === true) {
    return `${data.get('timeFrom')} - ${data.get('timeTill')}`;
  }
  return '';
};

class ExpandHourList extends React.Component<Props, State> {
  state = {
    show: false,
  };

  handleToggle = (e: Object) => {
    this.setState({ show: !this.state.show });
    e.stopPropagation();
  };

  render() {
    const { data, date } = this.props;
    if (!data || data.size === 0) {
      return null;
    }
    const today = moment(date).tz('America/Toronto');
    const todayWeekIndex = (today.day() - 1) % 7;
    const todayData = data.get(todayWeekIndex);

    const todayString = today.format('YYYY-MM-DD');
    const startDate = moment.tz(
      `${todayString} ${todayData.get('timeFrom')}`,
      'YYYY-MM-DD HH:mm A',
      'America/Toronto'
    );
    const endDate = moment.tz(
      `${todayString} ${todayData.get('timeTill')}`,
      'YYYY-MM-DD HH:mm A',
      'America/Toronto'
    );

    const businessOpened = today.isBetween(startDate, endDate);

    let status = 'Closed Now';
    let statusClass = 'expandHourList__status--danger';
    if (todayData && todayData.get('isActive') === true && businessOpened) {
      status = 'Open Now';
      statusClass = 'expandHourList__status--success';
    }

    return (
      <div className="expandHourList">
        <div
          className="expandHourList__labelWrapper"
          onClick={this.handleToggle}
          role="button"
        >
          <div className="expandHourList__label">Hours:</div>
          <div className={cx('expandHourList__status', statusClass)}>
            {status}
          </div>
          <div className="expandHourList__icon">
            <Icon glyph={this.state.show ? ChevronUp : ChevronDown} size={12} />
          </div>
        </div>
        {this.state.show && (
          <div className="expandHourList__dropdown">
            {data &&
              data.entrySeq().map(([key, value]) => {
                if (value.get('isActive') === false) {
                  return null;
                }
                return (
                  <div className="expandHourList__dropdownWrapper" key={key}>
                    <div className="expandHourList__label">&nbsp;</div>
                    <div
                      className={cx('expandHourList__status', {
                        'expandHourList__status--active':
                          key === todayWeekIndex,
                      })}
                    >
                      {weekNames[key]}
                    </div>
                    <div className="expandHourList__time">
                      {getTimeString(value)}
                    </div>
                  </div>
                );
              })}
          </div>
        )}
      </div>
    );
  }
}

export default ExpandHourList;
