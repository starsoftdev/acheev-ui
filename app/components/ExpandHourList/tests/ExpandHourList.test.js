import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import DEFAULT_HOURS_OF_OPERATION from 'enum/HoursOfOperation';

import ExpandHourList from 'components/ExpandHourList';

const data = fromJS(DEFAULT_HOURS_OF_OPERATION);

describe('<ExpandHourList />', () => {
  it("shouldn't throw error if no props are specified", () => {
    try {
      shallow(<ExpandHourList />);
    } catch (error) {
      expect(error).toBeUndefined();
    }
  });

  it('dropdown should be hidden by default', () => {
    const renderedComponent = shallow(<ExpandHourList data={data} />);
    expect(renderedComponent.find('.expandHourList__dropdown').length).toEqual(
      0
    );
  });

  it('should render dropdown when labelWrapper is clicked', () => {
    const renderedComponent = shallow(<ExpandHourList data={data} />);
    renderedComponent
      .find('.expandHourList__labelWrapper')
      .simulate('click', { stopPropagation: () => {} });
    expect(renderedComponent.find('.expandHourList__dropdown').length).toEqual(
      1
    );
  });

  it('should be open', () => {
    const component = create(
      <ExpandHourList date="2017-12-07T16:47-05:00" data={data} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be closed', () => {
    const component = create(
      <ExpandHourList date="2017-12-07T17:01-05:00" data={data} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
