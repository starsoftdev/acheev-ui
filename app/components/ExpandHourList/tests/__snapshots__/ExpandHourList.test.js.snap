// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpandHourList /> should be closed 1`] = `
<div
  className="expandHourList"
>
  <div
    className="expandHourList__labelWrapper"
    onClick={[Function]}
    role="button"
  >
    <div
      className="expandHourList__label"
    >
      Hours:
    </div>
    <div
      className="expandHourList__status expandHourList__status--danger"
    >
      Closed Now
    </div>
    <div
      className="expandHourList__icon"
    >
      <svg
        height={12}
        width={12}
      >
        <use
          xlinkHref="#undefined"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`<ExpandHourList /> should be open 1`] = `
<div
  className="expandHourList"
>
  <div
    className="expandHourList__labelWrapper"
    onClick={[Function]}
    role="button"
  >
    <div
      className="expandHourList__label"
    >
      Hours:
    </div>
    <div
      className="expandHourList__status expandHourList__status--success"
    >
      Open Now
    </div>
    <div
      className="expandHourList__icon"
    >
      <svg
        height={12}
        width={12}
      >
        <use
          xlinkHref="#undefined"
        />
      </svg>
    </div>
  </div>
</div>
`;
